### ISSUE TITLE
Gas estimation API is inconsistent

### ISSUE THREAD

**Issue by trruckerfling** [filecoin-project/fil-b#89]
**Description:** 
Gas estimation API for FEVM is inconsistent (most likely due to Filecoin's [deferred execution model](https://docs.filecoin.io/smart-contracts/filecoin-evm-runtime/difference-with-ethereum#deferred-execution-model)) and delivering poor builder experience: 
- returns low gas estimates and causes txn to time out and fail
- beyond 20% of gas overestimates, are burned and not refunded [(OP)](https://filecoinproject.slack.com/archives/C0745JU645A/p1732808561634869). 

**Proposed solution:**
- [gas benchmarking, suggested by steb](https://github.com/filecoin-project/ref-fvm/issues/1265) needs triage as it was raised in 2022
- Make changes to `eth_estimategas` (need triage but this seems high effort)
- Provide gas config examples or repos as [suggested by eva](https://filecoinproject.slack.com/archives/C0745JU645A/p1732875697226229?thread_ts=1732808561.634869&cid=C0745JU645A), for builders to reference

**Comment by Stebalien** [comment]
> most likely due to Filecoin's [deferred execution model](https://docs.filecoin.io/smart-contracts/filecoin-evm-runtime/difference-with-ethereum#deferred-execution-model).

This shouldn't affect gas estimation (significantly, at least). It could be that we're estimating gas at the wrong tipset, but any gas estimation that relies on a message being executed in a specific tipset is going to be flaky anyways.

> returns low gas estimates and causes txn to time out and fail

I assume you mean "fail with out of gas"? Messages can also "time out" (not be included fast enough for metamask) if the fee cap is too low. That's my guess given the recent high(er) gas prices but someone would need to confirm what "timed out" means in this context.

Our gas estimation is also likely nosier than in the EVM because of the underlying data model for FEVM smart contracts. In FEVM, the price of storing data in a smart contract varies with how much data is stored (as the smart contract's state grows) where it's a fixed cost per slot in the EVM. For new contracts with little state, this variability could easily be an issue.

> beyond 20% of gas overestimates, are burned and not refunded

Gas is refunded on a sliding scale. The more gas is over estimated, the less is refunded. 10% overestimation is "free". At 20%, we'll burn 10% of the overestimation (i.e., 2% of the total gas).

**Comment by trruckerfling** [comment]
thanks @Stebalien! Typical builder behaviour is to overestimate and set higher gas limits, to increase success of txn. So either improving gas estimation accuracy or full refund of overestimates will improve their experience. Eva raised a point ([here](https://filecoinproject.slack.com/archives/C0745JU645A/p1736428773803389?thread_ts=1736416313.584709&cid=C0745JU645A) and [here](https://filecoinproject.slack.com/archives/C0745JU645A/p1732808561634869)) that devs are used to getting full refunds and might opt for other chains if Filecoin burns above 10% of overestimate. 

Is there anything we can do to give refunds on overestimation above 10%?

**Comment by trruckerfling** [comment]
> would need to confirm what "timed out" means in this context

an nft mint and transfer failed because the gas estimation return very low gas limit, then tx execution failed of outOfGas error.

**Comment by trruckerfling** [comment]
in tackling accuracy of gas estimation API (builder feedback is that it's inconsistent), any ideas @Stebalien @snissn ? 

> It could be that we're estimating gas at the wrong tipset

could we test for this and correct if it's the case? 

From the proposed solutions, I looked up some gas optimization repos (e.g. [EVM Opcode Gas Costs](https://github.com/djrtwo/evm-opcode-gas-costs?tab=readme-ov-file) and a [simple guide](https://blog.web3labs.com/optimizing-gas-in-solidity-smart-contracts) to optimizing gas costs and it seems like a viable stopgap solution, although not actually solving the root issue.

**Comment by Stebalien** [comment]
> Eva raised a point ([here](https://filecoinproject.slack.com/archives/C0745JU645A/p1736428773803389?thread_ts=1736416313.584709&cid=C0745JU645A) and [here](https://filecoinproject.slack.com/archives/C0745JU645A/p1732808561634869)) that devs are used to getting full refunds and might opt for other chains if Filecoin burns above 10% of overestimate.

> Is there anything we can do to give refunds on overestimation above 10%?

The limit could be changed from, e.g., 10% to 20%. Or the rate at which it is burned could be changed. But we can't get rid of the overestimation burn unless we also get rid of tipsets or make significant changes to our execution model.

Basically, in Filecoin, we don't know how much gas a transaction will use until the next block is executed (until the block containing a transaction is included in a tipset). So we have to assume that it will use up to the specified limit. Over-estimating gas by too much _prevents_ others from using that same gas because the estimated gas counts against the block gas limit.

This isn't the case for chains without tipsets because they can count the _used_ gas against the block limit because they know how much gas will be used when the block is produced.

See https://filecoin.io/blog/posts/eip-1559-in-filecoin/#eip-1559-in-filecoin.

**Comment by Stebalien** [comment]
> > would need to confirm what "timed out" means in this context
>
> an nft mint and transfer failed because the gas estimation return very low gas limit, then tx execution failed of outOfGas error.

Are we sure they were using a _Filecoin_ node to estimate their gas and not relying on some external Ethereum-related tool?

**Comment by Stebalien** [comment]
>> It could be that we're estimating gas at the wrong tipset
>
> could we test for this and correct if it's the case?

In general, this shouldn't make much of a difference. Also, messages aren't guaranteed to be included in the next tipset so even fixing any off-by-ones here won't actually fix the issue.

My _guess_ is what I posted in https://github.com/filecoin-project/fil-b/issues/89#issuecomment-2583322604 if the estimates are actually "very low". But I don't have enough information to know that for sure.

> From the proposed solutions, I looked up some gas optimization repos (e.g. [EVM Opcode Gas Costs](https://github.com/djrtwo/evm-opcode-gas-costs?tab=readme-ov-file) and a [simple guide](https://blog.web3labs.com/optimizing-gas-in-solidity-smart-contracts) to optimizing gas costs and it seems like a viable stopgap solution, although not actually solving the root issue.

Not relevant to Filecoin because we don't use the Ethereum Gas model.

**Comment by Stebalien** [comment]
I'd say the next steps is make sure the correct gas estimation APIs are being used. I know that a lot of tools will use third-party services and/or fallback on hard-coded gas limits if they fail to estimate the gas. Knowing more about the exact situation would really help:

- Are these messages sent to a "new" contract without much state?
- What was the estimated gas, what was the actual gas.

Etc.

If our gas estimations tend to be off by 20% instead of, say, 10%, we can consider changing the cutoff where we start charging for overestimation. But we have to be careful here because it'll cut into chain bandwidth if we don't _also_ change the block gas limit.

**Comment by torchmike** [comment]
@trruckerfling are there concrete examples for us to investigate?

**Comment by snissn** [comment]
@trruckerfling  are there concrete examples for us to investigate?

**Comment by trruckerfling** [comment]
yeah, asking on [slack](https://filecoinproject.slack.com/archives/C0745JU645A/p1736796707380719?thread_ts=1736539557.917669&cid=C0745JU645A) to do a deep dive with concrete cases

**Comment by BigLep** [comment]
Per https://filecoinproject.slack.com/archives/C0745JU645A/p1738287174203539?thread_ts=1736539557.917669&cid=C0745JU645A, there is a proposal to close this.

There is also good discussion and analysis by @snissn in https://filecoinproject.slack.com/archives/C0745JU645A/p1738191717821889?thread_ts=1736539557.917669&cid=C0745JU645A about how Filecoin actually propagates that the failure in some cases is due to the contract reverting ([example](https://calibration.filfox.info/en/message/0x0f0dbfa9b5be461cb4a11e10a41a2ce7202e7efce1965c444d8c00947e836adc?t=4)).  sepola network does not.  (@snissn your call on how much you want to move your analysis/findings here so it's publicly discoverable/referencable for the future.)

The UX/DX is still understandably confusing though as evidenced by how even Filecoin builders are getting triggered by the "gas estimated failed" message and missing the underlying "contract reverted".  I personally don' t know how to improve that.   Maybe at the minimum there should be a document for the community around "If you see a 'Gas estimation failed' message be sure to first check X..."

**Comment by trruckerfling** [comment]
Yes let's close this since gas estimation is actually NOT inconsistent. Moving this to #98
